vpr/src/server/pathhelper.cpp

```diff
@@ -0,0 +1,117 @@
+        counter++;
+    }
+}
+
+/** 
+ * @brief helper function to calculate the setup critical path with specified parameters.
+ */
+static CritPathsResult generate_setup_timing_report(const SetupTimingInfo& timing_info, 
+                                                    const AnalysisDelayCalculator& delay_calc, 
+                                                    const t_analysis_opts& analysis_opts, 
+                                                    bool is_flat) {
+    auto& timing_ctx = g_vpr_ctx.timing();
+    auto& atom_ctx = g_vpr_ctx.atom();
+
+    VprTimingGraphResolver resolver(atom_ctx.nlist, atom_ctx.lookup, *timing_ctx.graph, delay_calc, is_flat);
+    resolver.set_detail_level(analysis_opts.timing_report_detail);
+
+    tatum::TimingReporter timing_reporter(resolver, *timing_ctx.graph, *timing_ctx.constraints);
+
+    std::vector<tatum::TimingPath> paths;
+    std::stringstream ss;
+    timing_reporter.report_timing_setup(paths, ss, *timing_info.setup_analyzer(), analysis_opts.timing_report_npaths);
+    collect_crit_path_metadata(ss, paths);
+    return CritPathsResult{paths, ss.str()};
+}
+
+/** 
+ * @brief helper function to calculate the hold critical path with specified parameters.
+ */
+static CritPathsResult generate_hold_timing_report(const HoldTimingInfo& timing_info, 
+                                                   const AnalysisDelayCalculator& delay_calc, 
+                                                   const t_analysis_opts& analysis_opts, 
+                                                   bool is_flat) {
+    auto& timing_ctx = g_vpr_ctx.timing();
+    auto& atom_ctx = g_vpr_ctx.atom();
+
+    VprTimingGraphResolver resolver(atom_ctx.nlist, atom_ctx.lookup, *timing_ctx.graph, delay_calc, is_flat);
+    resolver.set_detail_level(analysis_opts.timing_report_detail);
+
+    tatum::TimingReporter timing_reporter(resolver, *timing_ctx.graph, *timing_ctx.constraints);
+
+    std::vector<tatum::TimingPath> paths;
+    std::stringstream ss;
+    timing_reporter.report_timing_hold(paths, ss, *timing_info.hold_analyzer(), analysis_opts.timing_report_npaths);
+    collect_crit_path_metadata(ss, paths);
+    return CritPathsResult{paths, ss.str()};
+}
+
+/** 
+ * @brief Unified helper function to calculate the critical path with specified parameters.
+ */
+CritPathsResult calcCriticalPath(const std::string& type, int critPathNum, e_timing_report_detail detailsLevel, bool is_flat_routing) 
+{
+    // shortcuts
+    auto& atom_ctx = g_vpr_ctx.atom();
+
+    //Load the net delays
+    const Netlist<>& router_net_list = is_flat_routing ? (const Netlist<>&)g_vpr_ctx.atom().nlist : (const Netlist<>&)g_vpr_ctx.clustering().clb_nlist;
+    const Netlist<>& net_list = router_net_list;
+
+    NetPinsMatrix<float> net_delay = make_net_pins_matrix<float>(net_list);
+    load_net_delay_from_routing(net_list,
+                                net_delay);
```

Comment:
@soheilshahrouz 
implemented in follow way:
1) server is initialized by server_init as usual, regardless of current VPR state and may accept tasks.
2) in order to execute server tasks we have to initialize timing_info and delay_calc structures and bind them to server_ctx, which takes place exactly at VPR Routing stage.
Note:
By this way we don't have to return empty CritPathsResult. The filled CritPathsResult will be replied as soon as VPR enters Routing stage. This will simplify logic of client too, by not generating few same requests without knowing when exactly server is able to send response.
Hope this solution is ok for you. If yes, please click "resolve conversation".

resolved https://github.com/verilog-to-routing/vtr-verilog-to-routing/pull/2526/commits/cf3c11edef546e13bd6012fea40d7434b69da428